ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs

```diff
@@ -519,71 +522,242 @@ data HardForkLedgerWarning xs =
     -- /before/ this lower bound, the node is misconfigured and will likely
     -- not work correctly. This should be taken care of as soon as possible
     -- (before the transition happens).
-  | HardForkLedgerWarningUnexpectedTransition EraParams EpochNo
+  | HardForkWarningTransitionMismatch (EraIndex xs) EraParams EpochNo
+
+    -- | Transition in the final era
+    --
+    -- The final era should never confirm any transitions
+  | HardForkWarningTransitionInFinalEra (EraIndex xs) EpochNo
+
+    -- | An already-confirmed transition got un-confirmed
+  | HardForkWarningTransitionUnconfirmed (EraIndex xs)
+
+    -- | An already-confirmed transition got changed
+  | HardForkWarningTransitionReconfirmed (EraIndex xs) (EraIndex xs) EpochNo EpochNo
```

Comment:
Document the arguments. I'm now guessing: transition from, to, was confirmed for epoch, is reconfirmed in this other epoch.

Reply:
Ok, wrong guess :) Added: "We record the indices of the era we are transitioning from and to, as well as the old and new 'EpochNo' of that transition, in that order.".